codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java

```diff
@@ -579,102 +332,91 @@ public boolean isEmpty() {
     }
 
     @Override
-    public Headers<T> add(Headers<T> headers) {
+    public Headers<T> add(Headers<? extends T> headers) {
         checkNotNull(headers, "headers");
-
-        add0(headers);
+        if (headers == this) {
+            throw new IllegalArgumentException("can't add to itself.");
+        }
+        for (Entry<? extends T, ? extends T> header : headers) {
+            add(header.getKey(), header.getValue());
+        }
         return this;
     }
 
     @Override
+    @SuppressWarnings("unchecked")
```

Comment:
@buchgr - Yes disregard.

I think since I originally commented the `(Iterator<Entry<T, T>>)` cast and `hasMultipleValues` logic may have been added....or I also commented in the wrong place :)